### ISSUE TITLE
Need for Speed: Make Accel-Sim FAST again

### ISSUE THREAD

**Issue by JRPan** [accel-sim/accel-sim-framework#284]
<img width="722" alt="Screenshot 2024-02-21 at 14 12 48" src="https://github.com/accel-sim/accel-sim-framework/assets/25518778/3d721655-3031-46b3-8835-ce447f621abd">

Timing profile from linux perf:

1. `cache_stats` are called **per-cycle** by Accel-Wattch. 
2. xbar advance iterates all input->output which is `O(n^2)` which n is number of nodes. This is crazy. 

These two functions add up to 2/3 of the simulation time. `28+21+17 = 66%`. 

Fix:

1. include a global cache stat. So, you don't need to check for each SM each cycle. This is already done in `dev-stream-stats` branch. But needs some final polishing before being merged. For now, disable it if Accel-Wattch is not in use
2. only check the nodes that have an input/output. Even though, in the worst case, every node has a package, and the function still needs to iterate through all nodes but I doubt how often that happens. Anyway the worst case is the same, but on average this should be much better.

**Comment by JRPan** [comment]
See https://github.com/accel-sim/gpgpu-sim_distribution/pull/67 for the fix.

**Comment by rodhuega** [comment]
Hi, I have been studying the fix. I have one suggestion and one question.

Let's start with the suggestion. Regarding the cache stats, the fix only works in case the execution is not using Accel-Wattch. However, many of us need that component. I propose to change those if that you have added to be:  `if(m_config.g_power_simulation_enabled && (((gpu_tot_sim_cycle + gpu_sim_cycle) + 1) % m_config.gpu_stat_sample_freq == 0))` instead of  `if (m_config.g_power_simulation_enabled)` . In that way, the fix is also working in case Accel-Wattch is enabled. Notice the need for +1 because the `mcpat_cycle` function is triggered after `gpu_sim_cycle++;.` I have checked the energy report, and it provides the same outputs.

Moving to the question. I have tested the fix of the ICNT, and provides me different results. For example, a testing kernel that I have used to study the fix requires 2 cycles more to finish the execution. I wonder if this is expected or even if it is a better ICNT model. So, I would like to know if this matters or not.

PD: Thank you for the fix, I really appreciated.

**Comment by JRPan** [comment]
Thanks! Your solution regarding cache stats is better. I'll include that and keep that when we are adding the global cache stats.

Regarding ICNT, I expected it to be the same. I only intended to reduce the loop count. Do you mean it takes 2 cycles more in total to finish the execution? Or does each inst take 2 cycles more? 
Anyway, I have to test it more thoroughly. That's why I marked the PR as a draft and want input from the community. 
Thanks again, I will test it more. 

A little side talk: are you the author of this paper https://arxiv.org/pdf/2401.10082.pdf? We really like this work and it would be great if we could merge this. Is this something you may be interested in?

**Comment by JRPan** [comment]
Actually, I think I accidentally fixed a bug. lol. Someone correct me if I am wrong

in the original code:
```
for each output node:
    for each input node:
        if input_buffer[0] can be transferred from input node to output node:
            remove package from input buffer and insert into output buffer
```

But imagine an extreme case: input buffer 0 has 10 packages for output nodes 0 to 10. So: 
`input_buffer[] = 0,1,2 ... 9`

Then all 10 requests can be served within one cycle, which is wrong in design **iSLIP with a single iteration**. Each input should only be able to send 1 request each cycle.

Now in my design each output node can only accept one package per cycle. 

Some background reading http://nms.lcs.mit.edu/6829-papers/islip-ton.pdf: 
<img width="585" alt="image" src="https://github.com/accel-sim/accel-sim-framework/assets/25518778/93c0bbd2-a8af-49ea-ad97-ef1ac2d254aa">

Again, I may be wrong. Inputs are highly welcomed.

**Comment by rodhuega** [comment]
> Regarding ICNT, I expected it to be the same. I only intended to reduce the loop count. Do you mean it takes 2 cycles more in total to finish the execution? Or does each inst take 2 cycles more? Anyway, I have to test it more thoroughly. That's why I marked the PR as a draft and want input from the community. Thanks again, I will test it more.

It is reporting 2 cycles more of the stat `gpu_sim_cycle` of the first kernel of Backprop that belongs to Rodinia 2. As an effect of that, many other stats were changing too. It was in my simulator version, but I suppose that is going to be similar in the current dev version as I have never changed the ICNT between SMs and memory partitions.

> Actually, I think I accidentally fixed a bug. lol. Someone correct me if I am wrong
> 
> in the original code:
> 
> ```
> for each output node:
>     for each input node:
>         if input_buffer[0] can be transferred from input node to output node:
>             remove package from input buffer and insert into output buffer
> ```
> 
> But imagine an extreme case: input buffer 0 has 10 packages for output nodes 0 to 10. So: `input_buffer[] = 0,1,2 ... 9`
> 
> Then all 10 requests can be served within one cycle, which is wrong in design **iSLIP with a single iteration**. Each input should only be able to send 1 request each cycle.
> 
> Now in my design each output node can only accept one package per cycle.
> 
> Some background reading http://nms.lcs.mit.edu/6829-papers/islip-ton.pdf: <img alt="image" width="585" src="https://private-user-images.githubusercontent.com/25518778/307449958-93c0bbd2-a8af-49ea-ad97-ef1ac2d254aa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg3MjUzMTEsIm5iZiI6MTcwODcyNTAxMSwicGF0aCI6Ii8yNTUxODc3OC8zMDc0NDk5NTgtOTNjMGJiZDItYThhZi00OWVhLWFkOTctZWYxYWMyZDI1NGFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIzVDIxNTAxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcxZDM0YTAxYzMzNWNjNWQ0ZjlhODkyMGVmMWIzNWZiZDM0ZWIxNTVkYzcxYjNiNmYwNTc5ZDY1MWY3NDZjNDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Jf1yOnIIZCrBswAZzqcLQqrGxDdTfMq1ZwikAErkAB0">
> 
> Again, I may be wrong. Inputs are highly welcomed.

I'm not expert in ICNT, but I agree that the algorithm is saying that only accepts one input grant each cycle. However, I'm not sure if I am seeing where that is prevented compared to the original code. I suppose that the prevention comes from the original body loop

```
for (unsigned i = 0; i < total_nodes; ++i) {
  if (Has_Buffer_Out(i, 1)) {
    for (unsigned j = 0; j < total_nodes; ++j) {
....
       if (grant_cycles_count == 1)
              next_node[i] = (++node_id % total_nodes);
```

to your new loop that goes over the set, but I'm not identifying where. I guess that is something related with `start_node`, but I'm not getting how it is prevented that all the `dest` can have the same `next_node[dest]` in a given cycle.

> A little side talk: are you the author of this paper https://arxiv.org/pdf/2401.10082.pdf? We really like this work and it would be great if we could merge this. Is this something you may be interested in?

Yes, I'm one of the authors. I will be interested in merging that, but not right now. That arxiv PDF was a work in progress that we presented in CAMS 2023 and it is unfinished. We can discuss merging it when it is finish.

**Comment by JRPan** [comment]
It's prevented when building the set. https://github.com/accel-sim/gpgpu-sim_distribution/pull/67/files#diff-87ad483e1fe95a02fd73cf3251edacacbd39639eafa43b93150042ca68a0b511R190
In my code:
```
for each input
    package = input_node.head()
    if has package:
        input_set.insert(input_node)
        output_set.insert(output_node)
```

So I'm only checking the head of the input buffer. And both sets are unique. Then later in the nested loop, each input node and output node can only be iterated once. This prevented the bug. 

Of course it can still be improved. The temp_node is redundant. But functional-wise this is the same.

**Comment by rodhuega** [comment]
Ok, understood. I agree with you, now it is more correct.

By the way, I think these two modifications can also be applied to Vulkan-sim. Which I think is even slower than Accel-sim. I suppose this will have a good impact in that simulator too.

**Comment by JRPan** [comment]
Good idea! 
Thanks.

**Comment by rodhuega** [comment]
Hi, I have tested your new version of the code, and produces different results compared to the other two versions. I mean, original has IPC A, intermediate IPC B, and the current version has IPC C. I don't know if you were awere of the changes in IPC between intermediate code that we were discussing some days ago and the current state.

**Comment by JRPan** [comment]
Thank you for pointing that out. 

Yes, it should have some minor effects. The intermediate version was half-fixed.
I would expect original IPC > intermediate IPC > current IPC. Is this correct? 

In the intermediate version, an input node can still transfer multiple packages in some corner cases. 

However, I did not have time to thoroughly test everything (correlate every benchmark in the original paper). Our cluster is scheduled for upgrade. 
For now, I do not recommend using this for any results that go into a paper. 

But again, thanks for the input. I'll provide a full correlation once this change is ready.

**Comment by rodhuega** [comment]
Ok, I understand.

Regarding the IPC performance between versions, I have the following comparison with Backprop of Rodinia 2:
Initial:
gpu_tot_ipc =     361.4865
gpu_tot_ipc =     291.3595


Intermediate:
gpu_tot_ipc =     361.7782
gpu_tot_ipc =     291.9215


Final:
gpu_tot_ipc =     361.3072
gpu_tot_ipc =     291.6887


Until the first kernel:
Intermediate > Initial > Final

In the end:
Intermediate > Final > Initial

In my personal opinion, I won't care if the IPC is greater or smaller. It will change depending on the kernel. Moreover, the change is very little. I care more about the fact that now the code is doing what is expected.

Thanks for fixing that.

**Comment by tgrogers** [comment]
@William-An - can you add your KIPS numbers to this issue?
Has anyone run a perf on Accel-Sim recently? We should see if the bottleneck has shifted.

**Comment by tgrogers** [comment]
Is this technically called an issue? @JRPan what is the github project lingo for this?

**Comment by JRPan** [comment]
"Feature Enhancement"? I mean it's not a bug, but can still be an issue. A slowdown issue.

**Comment by tgrogers** [comment]
ok yeah feature enhancement works.

**Comment by tgrogers** [comment]
Also - which app were you running the original perf on?

**Comment by JRPan** [comment]
I don't really remember. But this should be universal.

**Comment by tgrogers** [comment]
Not necessarily.
Memory bound apps and compute-bound apps will spend more time in different sub-systems.
It would be cool to include a perf report as part of the regressions :)

**Comment by William-An** [comment]
I will post them once the PR runs are done.

> [@William-An](https://github.com/William-An) - can you add your KIPS numbers to this issue? Has anyone run a perf on Accel-Sim recently? We should see if the bottleneck has shifted.

**Comment by William-An** [comment]
Meanwhile, I am looking at our action runner, and it is interesting gpgpu-sim will have different run time based on different configs (PTX simulation only though over the past month):

<img width="897" alt="Image" src="https://github.com/user-attachments/assets/60f3110b-dec8-47e3-9f12-6ef0f1465263" />

**Comment by tgrogers** [comment]
Yeah, this is not surprising. Larger machines have more things to tick.
Since GPGPU-Sim is not event driven, but tick-driven, every level of every SM and memory partition is run on every simulated cycle - even when nothing happens.

**Comment by tgrogers** [comment]
This effect will be most pronounced in workloads that do not fill the machine - since in larger workloads, there will be more work/cycle but less cycles. In small workloads, you aren't using all the parallelism of the machine so the number of cycles stays relatively constant between big machine and little machine, but the sim time does not

**Comment by William-An** [comment]
# Result
I finished the run using the same GEMM kernel as with the compressed vs uncompressed case. On average, there is ~10% speedup after this patch:

<img width="494" alt="Image" src="https://github.com/user-attachments/assets/bf30eb47-78a5-423a-85ab-43404ba515d8" />

## Minor Caveat

It should be noted the baseline is not the head of the dev branch. The difference between the baseline and the patched gpgpusim is shown here: https://github.com/accel-sim/gpgpu-sim_distribution/compare/3844f7559e69f206a80ae9b970b19e7e609e949e...beb4cc4, which should not undermine the validity of the result as most changes are to the non-code part. The only part that is related to performance but not in the baseline is [this](https://github.com/accel-sim/gpgpu-sim_distribution/commit/45caf76587642b7d5dcd8b0992c55114c9ced35e), relating to the power model that is not enabled during the benchmarking.

## Run config
### Test env and workload setup:
- CUDA: 11.7
- GEMM:
  - CUTLASS 2048x2048x2048 FP32
  - Warp shape: 2x2x1
  - Threadblock shape: 128x128x8
- Run with compressed trace 3 times in series for each host

### Test commits
- baseline (compressed trace): https://github.com/accel-sim/gpgpu-sim_distribution/commit/3844f7559e69f206a80ae9b970b19e7e609e949e
- Fast patch: https://github.com/accel-sim/gpgpu-sim_distribution/commit/beb4cc4

**Comment by William-An** [comment]
There is an archive paper on using OpenMP to parallelize the SM cycle function of gpgpusim: https://arxiv.org/abs/2502.14691
